torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,365 @@
+        self._per_optimizer_states = defaultdict(dict)
+
+    def get_scale(self):
+        """
+        Returns:
+            The scale factor (a single-element :class:`torch.cuda.FloatTensor`), or 1.0 if scaling is disabled.
+
+        .. note::
+            :meth:`get_scale` alone does not incur a CPU-GPU sync, but if you wish to print the scale Tensor, or
+            inspect its value on the CPU by calling ``.item()``, that will incur a sync.
+        """
+        return self._scale if self._enabled else 1.0
+
+    def get_growth_factor(self):
+        r"""
+        Returns:
+            A Python float containing the scale growth factor.
+        """
+        return self._growth_factor
+
+    def set_growth_factor(self, new_factor):
+        r"""
+        Arguments:
+            new_scale (float):  Value to use as the new scale growth factor.
+        """
+        self._growth_factor = new_factor
+
+    def get_backoff_factor(self):
+        r"""
+        Returns:
+            A Python float containing the scale backoff factor.
+        """
+        return self._backoff_factor
+
+    def set_backoff_factor(self, new_factor):
+        r"""
+        Arguments:
+            new_scale (float):  Value to use as the new scale backoff factor.
+        """
+        self._backoff_factor = new_factor
+
+    def is_enabled(self):
+        r"""
+        Returns:
+            A bool indicating whether this instance is enabled.
+        """
+        return self._enabled
+
+    def state_dict(self):
+        r"""
+        Returns the state of the scaler as a :class:`dict`.  It contains three entries:
+
+        * ``"scale"`` - a :class:`torch.cuda.FloatTensor` containing the current scale
+        * ``"growth_factor"`` - a Python float containing the current growth factor
+        * ``"backoff_factor"`` - a Python float containing the current backoff factor
+
+        If this instance is not enabled, returns an empty dict.
+        """
+        return {"scale": self._scale,
+                "growth_factor": self._growth_factor,
+                "backoff_factor": self._backoff_factor} if self._enabled else {}
+
+    def load_state_dict(self, state_dict):
+        r"""
+        Loads the scaler state.  If this instance is disabled, :meth:`load_state_dict` is a no-op.
+
+        Arguments:
+           state_dict(dict): scaler state.  Should be an object returned from a call to :meth:`state_dict`.
+        """
+        if not self._enabled:
+            return
+
+        if len(state_dict) == 0:
+            raise RuntimeError("The source state dict is empty, possibly because it was saved "
+                               "from a disabled instance of AmpScaler.")
+
+        self._scale.copy_(state_dict["scale"])
+        self._growth_factor = state_dict["growth_factor"]
+        self._backoff_factor = state_dict["backoff_factor"]
+
+    def _check_inf_per_device(self, optimizer):
```

Comment:
this helper is not used anywhere?

Reply:
Not in any of the documented examples, but it is necessary for stochastic rounding optimizers, where you'd like to check the FP16 grads for infs/nans without unscaling.